File: php/convert_sorted_array_to_binary_search_tree.php
<?php

// 108.Convert Sorted Array to Binary Search Tree

// Given an integer array nums where the elements are sorted in ascending order, convert it to a height-balanced binary search tree.
//
//A height-balanced binary tree is a binary tree in which the depth of the two subtrees of every node never differs by more than one.

//  Definition for a binary tree node.
class TreeNode {
  public $val = null;
  public $left = null;
  public $right = null;
  function __construct($val = 0, $left = null, $right = null) {
      $this->val = $val;
      $this->left = $left;
      $this->right = $right;
  }
}

class Solution {

    /**
     * @param Integer[] $nums
     * @return TreeNode
     */
    function sortedArrayToBST($nums) {

        return $this->helper($nums,0,count($nums) - 1);
    }

    function helper($nums,$left,$right){
        if($left > $right){
            return null;
        }
        // 二叉树为有序数列，且为高度平衡二叉搜索树，选取中位数为根节点
        $mid = floor(($left + $right ) / 2);
        $root = new TreeNode($nums[$mid]);
        $root->left = $this->helper($nums,$left,$mid - 1);
        $root->right = $this->helper($nums,$mid + 1,$right);
        return $root;
    }

    function test($nums){
        // 生成的二叉树
        $root = $this->sortedArrayToBST($nums);
        var_dump($root);
        $mem = [];
        $this->traverse($root,$mem);
        // 生成的二叉树和原有的值是否相等
        foreach ($mem as $ko => $item) {
            if($nums[$ko] != $item){
                echo sprintf("test failed:element value:%d,target value:%d",$item,$nums[$ko]);
                return;
            }
        }
        echo "test success";
    }

    function traverse($root,&$mem){
        if($root == null){
            return null;
        }
        // 中序遍历
        $this->traverse($root->left,$mem);
        $mem[] = $root->val;
        $this->traverse($root->right,$mem);
        return $mem;
    }
}
$nums = [-10,-3,0,5,9];
$nums = [1,3];
$obj = new Solution();
$obj->test($nums);

File: go/convert_sorted_array_to_binary_search_tree.go
package main

import (
	"leet_code/tree"
	"log"
)

// 108.Convert Sorted Array to Binary Search Tree

// Given an integer array nums where the elements are sorted in ascending order, convert it to a height-balanced binary search tree.
//
//A height-balanced binary tree is a binary tree in which the depth of the two subtrees of every node never differs by more than one.


// 二叉搜索树中序遍历是升序序列，题目给出的是升序序列，是中序遍历结果
// 高度平衡的二叉搜索树，可以选择中间数据作为根节点，这样分给左右子树的数字个数相同或相差1
// 使得保持树平衡


var arr = []int{-10,-3,0,5,9}

var arr1 = []int{1,3}
func main()  {
	testSortedArrayToBST(arr1)
	//sortedArrayToBST(arr)
}


func sortedArrayToBST(nums []int) *tree.TreeNode {

	return helper(nums,0,len(nums) - 1)
}

func helper(nums []int,left,right int) *tree.TreeNode {
	if left > right {
		return nil
	}
	// 选取中位数
	mid :=  (left + right) / 2
	root := &tree.TreeNode{Val: nums[mid]}
	root.Left = helper(nums,left,mid - 1)
	root.Right = helper(nums,mid + 1,right)
	return root
}


func testSortedArrayToBST(nums []int) {
	root := sortedArrayToBST(nums)
	// 中序遍历应该是有序的
	mem := make([]int,0)
	ret := testTraverse6(root,&mem)
	// 和原数组元素是否相等
	for i,v := range *ret {
		if nums[i] != v {
			log.Printf("test error,element value :%d,target value:%d",v,nums[i])
			break
		}
	}
	log.Println("test success")
}


func testTraverse6(root *tree.TreeNode,ret *[]int) *[]int {
	if root == nil {
		return nil
	}
	testTraverse6(root.Left,ret)
	*ret = append(*ret,root.Val)
	testTraverse6(root.Right,ret)
	return ret
}